### PR TITLE
Enable a narrative to source multiple builds

### PR DESCRIPTION
**Pull request by eharkins**
### Description of proposed changes    
This builds on an existing contribution #1071 (thanks to the amazing work of @salvatore-fxpig 
 !) to address issue #1050, i.e. enabling a narrative to source multiple builds (datasets). It mostly works, but needs more testing.

**Some issues which might prevent this from getting merged** (and I am still working to reproduce and define so that we can address them):
1. There are a number of existing narratives where there is [a typo in one of the slides specifying the dataset](https://github.com/nextstrain/narratives/commit/c67986b3af2662bb3c09b4957838718dc7459718#diff-bd4829762586f75a51b1af9b5e34df52R179) or they [don't specify the build in exactly the same way](https://github.com/nextstrain/narratives/commit/d5ab4d15beccb4e0591f21b37ee79af35621dc05#diff-a75c26c5bef407811d683d2a18d1d007R34-R47). This wasn't an issue before, since having any of the datasets be different from the first slide doesn't have any effect, since that's the only one we used to load the dataset. With these changes, we fetch all the unique urls from all slides in a narrative, which might break such narratives as the ones I mentioned. @jameshadfield said:
> All of the narratives we control (e.g. github.com/nextstrain/narratives) we can write a small script to flag where the builds differ (they should have all been the same so far). It was spelled out in the original spec that the dataset should change, so if there are problems then people will have to fix their narrative files*
> \* if the fetch fails for a dataset change, we should fall-back to the currently-in-use dataset & throw up a notification. This’d be nice, but not essential.
2. Ensuring traditional narratives are not broken on mobile.
3. Ensuring that we implement componentWillUnmount() methods which remove listeners set up in any of the react components which might be displayed and then unmounted when changing slides (and datasets) in a narrative. This is to prevent memory leaks or updates to unmounted components (this may or may not be an issue currently with the Entropy component - componentWillUnmount still not implemented, see the diff).
4. I also am noticing a difference in some deme sizes between my branch and master on [a given ncov narrative](https://nextstrain.org/narratives/ncov/sit-rep/2020-04-03?f_region=Africa&n=4) I am using to test that the existing narrative behavior is unchanged - so will want to fix that as well assuming it's both wrong and introduced by these changes.

**Some goals it has inspired**:
1. The way we rerender all the main components when changing datasets in a narrative is with the [react key prop](https://reactjs.org/docs/lists-and-keys.html#keys). This makes it so we can avoid having to explicitly cover tons of cases of comparing props to guess if we should rerender a given component. This seems quite powerful and sets a cool precedent for other places where we'd like to do this in the app.
2. See below about a good idea James had for testing narratives.

**Some TODO items that can become separate issues** (i.e. don't prevent this from being merged if we don't mind these imperfections), which we will still appreciate any help with cc @salvatore-fxpig:
1. Ensuring multiple dataset narratives work on mobile.
2. Improve dataset fetching performance strategy (see TODOs in the diff)
3. Allow frequencies data to be fetched and displayed in narratives

### Testing
I have been testing with [this example narrative](https://nextstrain.org/narratives/test/multiple-datasets) created by @jameshadfield and extending it as necessary. As James pointed out, maintaining a huge narrative (or set of narratives) that runs through many combinations of dataset and queries (and thus app states) could be immensely helpful in testing narratives manually as well as with automated tests. I will create an issue as a stub for this and as a place for discussion of how we want to use that big test narrative.

### CODE REVIEW

File: src/components/narrative/index.js [hunk 1]
```diff
@@ -52,15 +52,16 @@ class Narrative extends React.Component {
...
-        // path: this.props.blocks[blockIdx].dataset, // not yet implemented properly
-        changeDataset: false,
+      const change = {
         query: queryString.parse(this.props.blocks[idx].query),
         queryToDisplay: {n: idx},
         push: true
-      }));
-
+      };
+      if (this.props.blocks[idx].dataset !== this.props.blocks[this.props.currentInFocusBlockIdx].dataset) {
```
Comment (jameshadfield) [inline src/components/narrative/index.js:60]: Is this the only section which needs to be transferred to the mobile narrative display for that to work?

File: src/actions/loadData.js [hunk 2]
```diff
@@ -239,6 +239,59 @@ export const loadSecondTree = (secondTreeUrl, firstTreeUrl) => async (dispatch,
...
+  let pathnameShouldBe;
+  const landingSlide = {
+    secondTreeDataset: false,
+    secondTreeName: false
+  };
+  // TODO:1050 A more performant fetching strategy would be fetching the dataset for the slide you land on,
+  // then fetching all the rest in the background so we can use them from the cache upon changing slides.
+  // Doing that presents the risk of a race case (if you change pages faster than a dataset can be fetched) so we are avoiding it for now.
+  // Instead we use Promise.all to ensure all the datasets are fetched before we render.
+  return Promise.all(blocks.map((block, i) => {
```
[the remark this answers is not in the document]
Comment (eharkins) [reply src/actions/loadData.js]: Added a commit that should fix this. Thanks for pointing it out!

**Comment by salvatore-fxpig** [comment]
Uuuh I was waiting for this to come up again!
Will give it a spin and see if I can help, the idea of using keys to forcing remounting is brilliant stuff. 👍 

PS: I think adding uniq from lodash just changed the other-vendors chunkhash, if you're already publishing somewhere with long-term caching I would suggest using `Set`, it should transpile just fine anyway.

**Comment by salvatore-fxpig** [comment]
I think the async stuff can be baked in pretty easily, something along these lines should do the trick
https://github.com/salvatore-fxpig/auspice/commit/a5afd508e59313903383f45a12de4f32035fcd37

**Comment by salvatore-fxpig** [comment]
Dumping some stuff here: #1167, hopefully makes it easier to play with
- getting rid of _.uniq sets the hash of the other-vendors bundle right again.
- Async loading
- Fallback to landing block, and from landing block to block[0], for missing dataset (with updated 404 result returned by server rather than generic 500)

File: src/reducers/jsonCache.js [hunk 3]
```diff
@@ -0,0 +1,20 @@
+import * as types from "../actions/types";
+
+/* the store to cache pre-loaded JSONS */
+
+const jsonCache = (state = {
+  jsons: {} /* object with dataset names as keys and loaded dataset / narrative jsons as values */
+}, action) => {
+  switch (action.type) {
+    case types.CACHE_JSONS:
```
[the remark this answers is not in the document]
Comment (eharkins) [reply src/reducers/jsonCache.js:9]: Makes sense! Just pushed a fix that seems to fix the bug with this behavior and only clear the cache on CLEAN_START

**Comment by eharkins** [comment]
Thanks @salvatore-fxpig !

> PS: I think adding uniq from lodash just changed the other-vendors chunkhash, if you're already publishing somewhere with long-term caching I would suggest using Set, it should transpile just fine anyway.

Makes sense. The reason I used _uniq instead of Sets was only because I wanted to use .map() on the unique list of datasets and not have to convert the Set back to an array to do that. And because when I used the jsons object keys to ensure no duplicate fetches, it didn't have that behavior because of a misunderstanding of promises on my part. I'm sure there is a better way though so I will check out 1167 and give it a test!

**Pull request merged**